### PR TITLE
WIP: Vault Integration

### PR DESCRIPTION
**Pull request by khos2ow**
## Description

This PR adds Vault integration:

- IKEV2 VPN implementation
- TODO

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [ ] I have added tests to cover my changes.
- [ ] All relevant new and existing integration tests have passed.
- [ ] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@khos2ow please rebase and re-open if still relevant

**Comment by skattoju4** [comment]
@DaanHoogland can we re-open this, attempting to revive this effort..

**Comment by rhtyd** [comment]
@skattoju3 Sure, or you may submit a new PR as this one may have a lot of conflicts.

**Comment by khos2ow** [comment]
@rhtyd: @skattoju4 has already rebased on master and fixed the conflicts on this branch, so it should be safe to reopen this one.

**Comment by rhtyd** [comment]
@khos2ow as the PR author you could reopen it yourself too, let me help reopen nonetheless.

**Comment by rhtyd** [comment]
@khos2ow @skattoju4 you need to rebase to latest master, PR says 2416 commits.

**Comment by khos2ow** [comment]
I think since I was not the person to close this, albeit being the author, I actually couldn't reopen myself. Thanks for reopening!

**Comment by skattoju4** [comment]
thanks for opening will fix the rebase..

**Comment by DaanHoogland** [comment]
@skattoju4 @khos2ow, please considder the PR #4904 to prevent conflicts later on.

**Comment by DaanHoogland** [comment]
> @DaanHoogland any significant change from #4904 that we require some code change from this current PR?
> 
> > #4904

never mind @pdion891 , it is a refactor of the ike vpn to allow for both v1 and v2, also it is merged and no conflicts arose here.

File: server/src/main/java/org/apache/cloudstack/pki/PkiManagerImpl.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import com.cloud.domain.Domain;
+import com.cloud.exception.RemoteAccessVpnException;
+import com.cloud.utils.component.ManagerBase;
+import com.cloud.utils.net.Ip;
+
+/**
+ * PKI Manager base class. This will work as a factory to construct Vault or Default
+ * implementation and pass through the API call to corresponding implementation.
+ *
+ * @author Khosrow Moossavi
```
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply server/src/main/java/org/apache/cloudstack/pki/PkiManagerImpl.java:39]: it happens but makes no sense legally (i think). I would be a marker for people to ask questions and spam Khosrow ;)

**Comment by pdion891** [comment]
What is preventing this PR to be merged at this point? I know we have to write doc on how to use it.

**Comment by DaanHoogland** [comment]
@pdion891 there are conflicts and @rhtyd :-1:'d this. nothing else ;)

**Comment by pdion891** [comment]
Hi @rhtyd, what did you -1  in this PR ?

**Comment by pdion891** [comment]
ok, found @rhtyd comment;

Basically we need to:
* create documentation.
* update UI fields to use existing one from the new UI. 
* make sure it still work fine on latest master branch.

Would something more be missing from this PR ?

**Comment by rhtyd** [comment]
Hi @pdion891 well I requested for some changes about two years ago and then did not see any engagement:
```
Please adapt plugin integration to CA framework, extend/refactor the CA framework as necessary. Several explicit ui elements can be removed, for example the download ca cert button already exists.
```

Here's what can happen:
- I don't understand where/how this feature is used (a design doc would be great), or what benefits it brings
- Does it require a separate/new dependency (external system), then it shouldn't be the default
- It can explore as a plugin implementation to the CA framework for cert generation/management
- IKEV2 is already supported in strongswan now afaik (the PR is old needs to be updated to latest `main` branch)

I came here from the ML thread, I don't see any openvpn use-case (unless that's a separate PR/thing, I did not see openvpn pkg/usage/configuration in this PR or pkg installed in systemvmtemplate).

**Comment by pdion891** [comment]
well this current PR intent is to leverage IKEv2 from current strongswan running on VR. this is not about OpenVPN.
where do we create Featurespec document nowaday, still in CloudStack wiki ?

**Comment by rhtyd** [comment]
@pdion891 ikev2 is supported now, see latest master. If you're extending certificate usage/management, I still think there's some work to be done on this old PR which is ~2yrs old. In its current state, it cannot be simply merged (see the conflict, sql upgrade path changes and rest of the code changes for example). Happy to help review, test, merge this PR if it can be revived in a state which is tested by the author(s) and reaches that point.

**Comment by rohityadavcloud** [comment]
Ping @khos2ow is this still valid or can be closed?

**Comment by sureshanaparti** [comment]
> Ping @khos2ow is this still valid or can be closed?

Hi @khos2ow Any update? If this is still valid, Can you fix the conflicts.

**Comment by khos2ow** [comment]
@rohityadavcloud @sureshanaparti unfortunately I'm not involved with the project and community anymore, and I'm not entirely up to date to the state of the project at master but I'm almost 100% sure folks at cloud.ca are still heavily relying on this feature. /cc @pdion891 @swill for clarification.

I handed off the project and this PR in particular to @skattoju4 which he rebased this PR some while ago and I think he should be able to follow up on it. I can rebase and fix the conflict as they are, but as I don't have the recent context and roadmap of the project I think it makes more sense for him to do that. Let me know either way.

**Comment by rohityadavcloud** [comment]
Thanks @khos2ow, given the PR hasn't made any activities in 3+yrs I'm closing this. Since you're not working on CloudStack anymore, pl re-open this or raise a fresh PR @pdion891 @swill in future.

**Pull request closed without merging**